validator/resource_value_pattern.rego

```diff
@@ -0,0 +1,151 @@
+# such that missing fields will trigger a violation.
+###########################
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	params := lib.get_constraint_params(constraint)
+	asset := input.asset
+
+	# asset type can be a scalar string or an array, only continue if the
+	# resource's asset type is one that should be looked at
+	# If asset types don't match exactly they will just be skipped
+	# potentially creating false positives.
+	# Ensure your assets types are 100% correct with negative testing.
+scalar_in(	asset.asset_type, params.asset_type)
+
+	field_name := params.field_name
+
+	rule_data := {
+		"is_required": lib.get_default(params, "required", true),
+		"has_field": has_field_by_path(asset.resource.data, field_name),
+		"has_pattern": lib.has_field(params, "pattern"),
+		"pattern": lib.get_default(params, "pattern", ""),
+		"value": get_default_by_path(asset.resource.data, field_name, ""),
+	}
+
+	# The rule is truthy so we ensure the asset is valid
+	# and only violate if it is not as part of the deny rule
+not is_valid(	rule_data)
+
+	message := sprintf("%v does not have a valid field named '%v' with a value '%v' matching pattern '%v'.", [
+		asset.name,
+		field_name,
+		rule_data.value,
+		rule_data.pattern,
+	])
+
+	metadata := {"resource": asset.name}
+}
+
+###########################
+# Rule Utilities
+###########################
+
+is_valid({
+	"is_required": is_required,
+	"has_field": has_field,
+	"has_pattern": has_pattern,
+	"pattern": pattern,
+	"value": field_value,
+}) {
+	is_required_field_valid(is_required, has_field)
+	is_pattern_valid(is_required, has_field, has_pattern, pattern, field_value)
+}
+
+is_required_field_valid(is_required, has_field) {
+	has_field == true
+}
+
+is_required_field_valid(is_required, has_field) {
+	is_required == false
+	has_field == false
+}
+
+is_pattern_valid(is_required, has_field, has_pattern, pattern, value) {
+	has_pattern == false
+}
+
+is_pattern_valid(is_required, has_field, has_pattern, pattern, value) {
+	has_field == false
+}
+
+is_pattern_valid(is_required, has_field, has_pattern, pattern, value) {
+	has_pattern == true
+	has_field == true
+	re_match(pattern, value)
+}
+
+###
+# Rule Library Functions
+###
+scalar_in(scalar, domain) {
+	is_string(domain)
+	scalar == domain
+}
+
+scalar_in(scalar, domain) {
+	is_array(domain)
+	domain[_] == scalar
+}
+
+scalar_in(scalar, domain) {
+	is_set(domain)
+	domain[_] == scalar
+}
+
+get_field_by_path(obj, path) = output {
+	split(path, ".", path_parts)
+	walk(obj, [path_parts, output])
+}
+
+# internal helper because walk has strange behavior
+# such that we need to wrap it in order to get
+# explicit true / false
+_has_field_by_path(obj, path) {
+	_ := get_field_by_path(obj, path)
+}
+
+has_field_by_path(obj, path) {
+	_has_field_by_path(obj, path)
+}
+
+has_field_by_path(obj, path) = false {
+	not _has_field_by_path(obj, path)
+}
+
+get_default_by_path(obj, path, _default) = output {
```

Comment:
what's the significance of the underscore in front of `default`? that seems inconsistent from other parameters?

Reply:
Its modeled identical to the existing rules in lib.  Should I change or keep them?

https://github.com/forseti-security/policy-library/blob/0dfa9432ea261e7605146d03a64a490e860e0327/lib/util.rego#L38

Reply:
ah, never mind, `default` is a keyword, and that's the reason for the underscore prefix.